src/sys/socket/mod.rs

```diff
@@ -369,99 +389,95 @@ pub struct CmsgSpace<T> {
+        // Get the data.
+        // Safe if cmsghdr points to valid data returned by recvmsg(2)
+        let cm = unsafe { Some(ControlMessageOwned::decode_from(self.cmsghdr))};
+        // Advance the internal pointer
+        // Safe if mhdr and cmsghdr point to valid data returned by recvmsg(2)
+        self.cmsghdr = unsafe{
+            CMSG_NXTHDR(self.mhdr as *const _, self.cmsghdr as *const _)
         };
-
-        let cmsg_len = cmsg.cmsg_len as usize;
-
-        // Advance our internal pointer.
-        let cmsg_data = &self.buf[cmsg_align(mem::size_of::<cmsghdr>())..cmsg_len];
-        self.buf = &self.buf[cmsg_align(cmsg_len)..];
-
-        // Safe if: `cmsg_data` contains the expected (amount of) content. This
-        // is verified by the kernel.
-        unsafe {
-            Some(ControlMessage::decode_from(cmsg, cmsg_data))
-        }
+        cm
     }
 }
 
-/// A type-safe wrapper around a single control message. More types may
-/// be added to this enum; do not exhaustively pattern-match it.
+/// A type-safe wrapper around a single control message, as used with
+/// [`recvmsg`](#fn.recvmsg).
+///
 /// [Further reading](http://man7.org/linux/man-pages/man3/cmsg.3.html)
+//  Nix version 0.13.0 and earlier used ControlMessage for both recvmsg and
+//  sendmsg.  However, on some platforms the messages returned by recvmsg may be
+//  unaligned.  ControlMessageOwned takes those messages by copy, obviating any
+//  alignment issues.
+//
+//  See https://github.com/nix-rust/nix/issues/999
 #[allow(missing_debug_implementations)]
-pub enum ControlMessage<'a> {
-    /// A message of type `SCM_RIGHTS`, containing an array of file
-    /// descriptors passed between processes.
-    ///
-    /// See the description in the "Ancillary messages" section of the
-    /// [unix(7) man page](http://man7.org/linux/man-pages/man7/unix.7.html).
-    ///
-    /// Using multiple `ScmRights` messages for a single `sendmsg` call isn't recommended since it
-    /// causes platform-dependent behaviour: It might swallow all but the first `ScmRights` message
-    /// or fail with `EINVAL`. Instead, you can put all fds to be passed into a single `ScmRights`
-    /// message.
-    ScmRights(&'a [RawFd]),
-    /// A message of type `SCM_CREDENTIALS`, containing the pid, uid and gid of
-    /// a process connected to the socket.
-    ///
-    /// This is similar to the socket option `SO_PEERCRED`, but requires a
-    /// process to explicitly send its credentials. A process running as root is
-    /// allowed to specify any credentials, while credentials sent by other
-    /// processes are verified by the kernel.
-    ///
-    /// For further information, please refer to the
-    /// [`unix(7)`](http://man7.org/linux/man-pages/man7/unix.7.html) man page.
-    // FIXME: When `#[repr(transparent)]` is stable, use it on `UnixCredentials`
-    // and put that in here instead of a raw ucred.
+pub enum ControlMessageOwned {
```

Comment:
Per [my previous comment](https://github.com/nix-rust/nix/issues/999#issuecomment-454601121), I think this is the wrong way to go, as it incurs avoidable heap allocation and copying.

Reply:
I chose to go this path because it's closer to backwards-compatible.  I tried the 100% backwards-compatible method first, but the code got very ugly very quick, and it involved even more copying.

Reply:
It seems to be that you've settled on the worst of both worlds: both a breaking change and inefficient.

Reply:
Can you suggest something more efficient that causes no more pain for existing users?  This solution, while not exactly backwards compatible, only requires a simple search and replace.

Reply:
I don't think the specific level of breakage makes sense as a line in the sand that cannot be crossed by a single step. That said, you could compromise by only using the flyweight pattern where heap allocation is otherwise necessary, i.e. in `ScmRights`, though this would make "Owned" no longer an accurate descriptor.

Reply:
But what's to say that `ScmRights`'s payload is aligned on all platforms that Nix will ever support?  I don't think that's guaranteed, especially if a user ever passes an unaligned msg_control field to recvmsg.

Reply:
That's why I said "flyweight pattern" and not "a slice", yes. Refer to the comment I linked for an example implementation.